File: python/binarytreeinordertraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        if not root:
            return res
        stack = []
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                t = stack.pop()
                res.append(t.val)
                root = t.right
        return res

File: cpp/binarytreeinordertraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if (!root) return vector<int>();
        stack<TreeNode*> stack;
        vector<int> res;
        while (root or !stack.empty()) {
            if (root) {
                stack.push(root);
                root = root->left;
            }
            else {
                auto target = stack.top();
                stack.pop();
                res.push_back(target->val);
                root = target->right;
            }
        }
        return res;
    }
};
